uncertainties/core.py

```diff
@@ -2766,6 +2779,22 @@ def std_dev(self, std_dev):
 
         self._std_dev = float(std_dev)
 
+    def __hash__(self):
+        """
+        Calculates the hash for any Variable object.
+        This simply calls the __hash__ method of `AffineScalarFunc` in most cases.
+        During initialization, the `_linear_part` and `_nominal_value` attributes does not exist.
+        Because of that, the id returned in case this method is called during initialization.
+        
+        Returns:
+            int: The hash of this object
+        """
+        # Otherwise, pickles loads does not work
+        if not hasattr(self, "_nominal_value"):
+            self._nominal_value = None
```

Comment:
I think the problem is, that pickle already generates the `_linear_part` object, before calling `__setstate__`.
Since the `_linear_part` always refers to the `Variable` it self, it cannot calculate the hash properly.

I think we can prevent that by deleting the self reference inside the `__getstate__` method. However, the derivative of `self` must not be one right? That means, we have to serialize the derivative in some way.

Reply:
I tried some stuff to handle this in the `__getstate__` and `__setstate__` methods.
But one single unit test is failing: 

```py
    # Corner case test: when an attribute is present both in __slots__
    # and in __dict__, it is first looked up from the slots
    # (references:
    # http://docs.python.org/2/reference/datamodel.html#invoking-descriptors,
    # http://stackoverflow.com/a/15139208/42973). As a consequence,
    # the pickling process must pickle the correct value (i.e., not
    # the value from __dict__):
    x = NewVariable_dict(3, 0.14)
    x._nominal_value = 'in slots'
    # Corner case: __dict__ key which is also a slot name (it is
    # shadowed by the corresponding slot, so this is very unusual,
    # though):
    x.__dict__['_nominal_value'] = 'in dict'
    # Additional __dict__ attribute:
    x.dict_attr = 'dict attribute'

    x_unpickled = pickle.loads(pickle.dumps(x))
```

I do not completely understand, what this test is for.
I think the problem might be the `__getstate__` method.
I will check this in more detail on Monday.

Reply:
That would only work if it is guaranteed, that the `_linear_part` of `AffineScalarFunction` does not contain a self reference. Is that the case?

Reply:
I think so. I think it is only `Variable` that has the self-reference since `LinearCombination` only holds `Variable`s, so a more complicated `AffineScalarFunc` instance would not put itself into the `LinearCombination`.

Reply:
I implemented the override of `__setstate__` and `__getstate__` for `Variable` and it seems to work. All test cases are passing on my local machine.